### PR TITLE
Optimize SendPacketAsync method for MQTTnet.AspNetCore

### PR DESCRIPTION
**Pull request by xljiulang**
1. [Avoid to call MqttPacketBuffer.Join() when SendPacketAsync](https://github.com/dotnet/MQTTnet/commit/a2127c68fd86692979031871bc51af969e8b6c7c)
2. [Safely cleanup the PacketFormatterAdapter](https://github.com/dotnet/MQTTnet/commit/df8e78a50b31124c361c7fdb0f2ca1e6c06542d9)

### CODE REVIEW

**Review by chkr1011 — CHANGES_REQUESTED** [review]

File: Source/MQTTnet.AspnetCore/MqttConnectionContext.cs [hunk 1]
```diff
@@ -169,20 +169,24 @@ public void ResetStatistics()
 
         public async Task SendPacketAsync(MqttPacket packet, CancellationToken cancellationToken)
         {
-            var formatter = PacketFormatterAdapter;
-            
             using (await _writerLock.EnterAsync(cancellationToken).ConfigureAwait(false))
             {
-                var buffer = formatter.Encode(packet);
-                var msg = buffer.Join().AsMemory();
```
[the remark this answers is not in the document]
Comment (xljiulang) [reply Source/MQTTnet.AspnetCore/MqttConnectionContext.cs:177]: We can detect whether the transmission is `mqtt over websocket` and then keep the join.

**Comment by xljiulang** [comment]
@chkr1011 
I've added a `readonly bool? _isOverWebSocket` field, we keep using `buffer.Json()` when it's value is true or null.

File: Source/MQTTnet.AspnetCore/MqttConnectionContext.cs [hunk 2]
```diff
@@ -8,28 +8,34 @@
...
+
         public MqttConnectionContext(MqttPacketFormatterAdapter packetFormatterAdapter, ConnectionContext connection)
         {
             PacketFormatterAdapter = packetFormatterAdapter ?? throw new ArgumentNullException(nameof(packetFormatterAdapter));
             Connection = connection ?? throw new ArgumentNullException(nameof(connection));
 
+            var feature = connection.Features.Get<IHttpContextFeature>();
+            if (feature != null)
+            {
+                _isOverWebSocket = feature.HttpContext.WebSockets.IsWebSocketRequest;
```
Comment (chkr1011) [inline Source/MQTTnet.AspnetCore/MqttConnectionContext.cs:36]: Just out of curiosity, there any chance that this will be not a web socket request?

Comment (xljiulang) [reply Source/MQTTnet.AspnetCore/MqttConnectionContext.cs:36]: Yes, `MqttConnectionHandler` has two usage scenarios, `UseMqtt` and `MapMqtt`, for `over tcp` and `over WebSocket`.

**Comment by chkr1011** [comment]
Thank you for your contribution. Good work 👍

**Pull request merged**